dataprep/clean/clean_duplication_utils.py

```diff
@@ -0,0 +1,323 @@
+"""
+Common functions and classes for the clean_duplication function.
+"""
+
+import string
+import unicodedata
+import collections
+import base64
+from operator import itemgetter
+from typing import List, Set, DefaultDict
+
+import pandas as pd
+from IPython.display import Javascript, display
+from metaphone import doublemetaphone
+
+
+class Clusterer:
+    """
+    Performs clustering methods on data.
+    """
+
+    clusters: pd.Series
+    _df: pd.DataFrame
+    _df_name: str
+    _col: str
+    _ngram: int
+    _radius: int
+    _block_size: int
+
+    def __init__(self, df: pd.DataFrame, col_name: str, df_name: str):
+        self.clusters = pd.Series()
+        self._df = df
+        self._df_name = df_name
+        self._col = col_name
+        self._ngram = 2
+        self._radius = 2
+        self._block_size = 6
+        self._df[self._col] = self._df[self._col].astype(str)
+
+    def cluster(self, cluster_method: str) -> None:
+        """
+        Create clusters using the given clustering method.
+        """
+        if cluster_method == "levenshtein":
+            self._nearest_neighbours_cluster()
+        else:
+            self._key_collision_cluster(cluster_method)
+
+    def _key_collision_cluster(self, cluster_method: str) -> None:
+        """
+        Create clusters using a key collision method.
+        Clusters are a Pandas Series of lists (each list represents a cluster),
+        each list contains tuples with the form (item, count).
+        """
+        key_funcs = {
+            "fingerprint": self._finger_print_key,
+            "ngram-fingerprint": self._ngram_finger_print_key,
+            "phonetic-fingerprint": self._phonetic_fingerprint_key,
+        }
+        key_func = key_funcs[cluster_method]
+        col = self._col
+        # get the count of each item in the dataframe and remove duplicates
+        counts = self._df.groupby(col).size()
```

Comment:
I think if this is the unique counts of values in a series, the API `self._df[col].value_counts(sort=False)` might be more appropriate here. I could be wrong though and maybe it's not what you want.

Reply:
`value_counts()` is definitely better. It was introduced in 1.1 though, are we still supporting pandas versions before that?

Reply:
Good point, I did not know that. Since we're already using `value_counts()` in EDA and Clean, I think we can update the dependency to support pandas versions on or after 1.1.

Reply:
Sounds good, I didn't know we were already using it.